runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -208,6 +209,55 @@ bool J9::TransformUtil::avoidFoldingInstanceField(
       "avoidFoldingInstanceField: symbol %p is never foldable (expected possibly foldable)\n",
       field);
 
+   if (fej9->isStable(field, cpIndex, owningMethod, comp) && !field->isFinal())
+      {
+      uintptr_t fieldAddress = object + fieldOffset;
+
+      TR::DataType loadType = field->getDataType();
+      switch (loadType)
+         {
+         case TR::Int32:
+            {
+            int32_t value = *(int32_t*)fieldAddress;
+            if (value == 0)
+               return true;
+            }
+            break;
+         case TR::Int64:
+            {
+            int64_t value = *(int64_t*)fieldAddress;
+            if (value == 0)
+               return true;
+            }
+            break;
+         case TR::Float:
+            {
+            float value = *(float*)fieldAddress;
+            if (value == 0.0)
+               return true;
+            }
+            break;
+         case TR::Double:
+            {
+            double value = *(double*)fieldAddress;
+            if (value == 0.0)
+               return true;
+            }
+            break;
```

Comment:
These comparisons `== 0.0` are true for negative zero as well, but negative zero is a different value from the default and so maybe it can be folded - I'm not sure of the precise rules around that for `@Stable`. It's probably just fine to leave this as it is with a comment, since refusing to fold is conservative. If anything is relying on folding negative zero for performance (which seems unlikely), this can be made more aggressive later

Also note that the comparison will be false when value is NaN, allowing NaN values to be folded. To me that seems very reasonable, but again I don't know the precise rules for `@Stable` floating-point numbers. If the situation for NaN is unclear, preventing folding would be more conservative. In either case, NaN may also be worth a comment

Reply:
The closest documentation I could find was: 

"Since all fields begin with a default value of null for references
(resp., zero for primitives), it follows that this annotation indicates
that the first non-null (resp., non-zero) value stored in the field
will never be changed."

So, yes, I assumed that NaN can be folded and not folding negative 0 should be safe. I was thinking that maybe most appropriate way is to cast to integral type of the same size and check for 0 then. Basically checking if all bits are 0s, which I think was the intent of Stable. On the other hand, not folding -0.0 is just a bit safer. 

I think you are right: a comment is needed. But it's ok to leave the code as is.